### PR TITLE
DFT wrapper supporting conventional DFT methods

### PR DESCRIPTION
**Pull request by wxj6000**
This PR
- supports some conventional DFT methods combining XC and dispersion correction/nonlocal correction, e.g wb97m-d3bj = wb97m-v + d3bj, b3lyp-d3bj = b3lyp + d3bj
- is backward compatible. No API changes for users.
- works through `mf=rks.RKS(mol, xc='wb97m-d3bj')` only, not `mf.xc=wb97m-d3bj`. `mf.xc` is defined as the same as before. `mf.xc` and `mf.disp` are used for customizing DFT methods.
- replaces `ni.libxc.is_nlc` with `ni.is_nlc` such that `ni.is_nlc` can be customized.

### CODE REVIEW

**Comment by wxj6000** [comment]
#2069

File: pyscf/dft/rks.py [hunk 1]
```diff
@@ -260,6 +260,25 @@ def define_xc_(ks, description, xctype='LDA', hyb=0, rsh=(0,0,0)):
...
+    xc = dft_method
+    disp = None
+
+    # special cases
+    if method_lower == 'wb97m-d3bj':
+        return 'wb97m-v', False, 'd3bj'
+    if method_lower == 'b97m-d3bj':
+        return 'b97m-v', False, 'd3bj'
+    if method_lower == 'wb97x-d3bj':
+        return 'wb97x-v', False, 'd3bj'
```
Comment (hebrewsnabla) [inline pyscf/dft/rks.py:275]: wb97x-d3 is also a special case.

**Comment by hebrewsnabla** [comment]
> works through mf=rks.RKS(mol, xc='wb97m-d3bj') only, not mf.xc=wb97m-d3bj.

could you add an example for those functionals?

File: pyscf/dft/rks.py [hunk 2]
```diff
@@ -260,6 +260,46 @@ def define_xc_(ks, description, xctype='LDA', hyb=0, rsh=(0,0,0)):
...
+DISP_VERSIONS = ['d3bj', 'd3zero', 'd3bjm', 'd3zerom', 'd3op', 'd4']
+
+def dft_method_parser(dft_method):
+    ''' conventional dft method ->
+    ( internal xc,
+      disable nlc,
+      dispersion version
+      with 3-body dispersion or not
+      '''
+    method_lower = dft_method.lower()
```
Comment (sunqm) [inline pyscf/dft/rks.py:273]: Please check if xc is a string and return early. The input xc can be a tuple of strings or other data types.

File: pyscf/dft/rks.py [hunk 3]
```diff
@@ -321,10 +361,13 @@ class KohnShamDFT:
     -76.415443079840458
     '''
 
-    _keys = {'xc', 'nlc', 'grids', 'nlcgrids', 'small_rho_cutoff'}
+    _keys = {'xc', 'nlc', 'grids', 'disp', 'nlcgrids', 'small_rho_cutoff'}
```
Comment (sunqm) [inline pyscf/dft/rks.py:364]: Add the key 'disp_with_3body'

File: pyscf/dft/rks.py [hunk 4]
```diff
@@ -339,6 +382,11 @@ def __init__(self, xc='LDA,VWN'):
 # don't modify the following attributes, they are not input options
         self._numint = numint.NumInt()
 
+        # disable nlc for certain xc such as wb97m-d3bj
+        if with_nlc is not None:
```
Comment (sunqm) [inline pyscf/dft/rks.py:386]: if one sets `nlc`, or `xc` after creating RKS object, the `_numint.nlc_coeff` and `self._numint.is_nlc` can never parse the xc correctly. Are there any other ways to control this?

**Comment by sunqm** [comment]
> > works through mf=rks.RKS(mol, xc='wb97m-d3bj') only, not mf.xc=wb97m-d3bj.
> 
> could you add an example for those functionals?

This is my suggestion to work so.

The additional parser in this PR can map these functionals to certain special combinations of other functionals when initializing DFT. If you manually set the `xc` attribute for a DFT object, that sounds you force the DFT to use the  specified functional. I think warnings can be issued in XC parser when these functionals are detected, and suggestions can be put in the warning message. But the DFT calculation should still use the specified xc functional.

File: pyscf/dft/rks.py (hunk 4 above)
Comment (wxj6000) [reply pyscf/dft/rks.py:386]: `nlc` is possibly disabled only when creating RSK object with certain functionals such as wb97m-d3bj. After the creation, setting nlc or xc does not change `self._numint.is_nlc`. The functionality of setting nlc or xc after creating RKS object is the same as before.

And one still can use `self._numint.libxc.is_nlc` and `self._numint.libxc.nlc_coeff` to parse xc code if `self._numint.is_nlc` is customized.

Comment (wxj6000) [reply pyscf/dft/rks.py:388]: Yes, it is better to customize the parser in those two functions.
https://github.com/wxj6000/pyscf/blob/dft_wrapper/pyscf/dft/numint.py#L2647
and 
https://github.com/wxj6000/pyscf/blob/dft_wrapper/pyscf/dft/numint.py#L2644
I will work on it later.

File: pyscf/dft/dft_parser.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        return 'wb97x-v', False, ('wb97x', 'd3bj', False)
+
+    # J. Chem. Theory Comput. 2013, 9, 1, 263–272
+    if method_lower in ['wb97x-d3']:
+        raise NotImplementedError('wb97x-d3 is not supported yet.')
+
+    if method_lower.endswith('-3c'):
+        raise NotImplementedError('*-3c methods are not supported yet.')
+
+    for d in DISP_VERSIONS:
```
Comment (sunqm) [inline pyscf/dft/dft_parser.py:63]: May need to consider a corner case like `wb97m-d3bjm`

**Comment by sunqm** [comment]
> > works through mf=rks.RKS(mol, xc='wb97m-d3bj') only, not mf.xc=wb97m-d3bj.
> 
> could you add an example for those functionals?

Should work in both cases with the latest updates.

File: pyscf/dft/dft_parser.py (hunk 5 above)
Comment (wxj6000) [reply pyscf/dft/dft_parser.py:63]: `wb97m-d3bjm` is not a well-defined xc functional. In this parser, `wb97m-d3bjm` will be interpreted as `wb97m` and `d3bjm`. `wb97m` is not a well-defined xc functional. I am assuming xc parser or libxc will raise an error.

**Pull request merged**